pkg/promtail/targets/journaltarget.go

```diff
@@ -0,0 +1,224 @@
+	journalPath := targetConfig.Path
+	if journalPath == "" {
+		journalPath = "/var/log/journal"
+	}
+
+	var err error
+	t.r, err = readerFunc(sdjournal.JournalReaderConfig{
+		Path:      journalPath,
+		Cursor:    position,
+		Since:     targetConfig.Since,
+		Formatter: t.formatter,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "creating journal reader")
+	}
+
+	go func() {
+		err := t.r.Follow(until, ioutil.Discard)
+		if err != nil && err != sdjournal.ErrExpired {
+			level.Error(t.logger).Log("msg", "received error during sdjournal follow", "err", err.Error())
+		}
+	}()
+
+	return t, nil
+}
+
+func (t *JournalTarget) formatter(entry *sdjournal.JournalEntry) (string, error) {
+	ts := time.Unix(0, int64(entry.RealtimeTimestamp)*int64(time.Microsecond))
+
+	msg, ok := entry.Fields["MESSAGE"]
+	if !ok {
+		level.Debug(t.logger).Log("msg", "received journal entry with no MESSAGE field")
+		return journalEmptyStr, nil
+	}
+	entryLabels := makeJournalFields(entry.Fields)
+
+	// Add constant labels
+	for k, v := range t.labels {
+		entryLabels[string(k)] = string(v)
+	}
+
+	processedLabels := relabel.Process(labels.FromMap(entryLabels), t.relabelConfig...)
+
+	processedLabelsMap := processedLabels.Map()
+	labels := make(model.LabelSet, len(processedLabelsMap))
+	for k, v := range processedLabelsMap {
+		if k[0:2] == "__" {
+			continue
+		}
+
+		labels[model.LabelName(k)] = model.LabelValue(v)
+	}
+	if len(labels) == 0 {
+		// No labels, drop journal entry
+		return journalEmptyStr, nil
+	}
+
+	t.positions.PutString(t.positionPath, entry.Cursor)
+	err := t.handler.Handle(labels, ts, msg)
+	return journalEmptyStr, err
+}
+
+// Type returns JournalTargetType.
+func (t *JournalTarget) Type() TargetType {
+	return JournalTargetType
+}
+
+// Ready indicates whether or not the journal is ready to be
+// read from.
+func (t *JournalTarget) Ready() bool {
+	return true
+}
+
+// TODO(rfratto): Perhaps the Target interface should remove DiscoveredLabels
+// and Labels and instead have a Dropped method.
+
+// DiscoveredLabels satisfies the Target interface. Returns nil for
+// JournalTarget as there are no discovered labels present as a
+// JournalTarget processes.
+func (t *JournalTarget) DiscoveredLabels() model.LabelSet {
+	return nil
+}
+
+// Labels satisfies the Target interface. Returns nil for JournalTarget
+// as there are is no guaranteed constant list of labels present as a
+// JournalTarget processes.
+func (t *JournalTarget) Labels() model.LabelSet {
+	return nil
+}
+
+// Details returns target-specific details (currently nil).
+func (t *JournalTarget) Details() interface{} {
+	return nil
+}
+
+// Stop shuts down the JournalTarget.
+func (t *JournalTarget) Stop() error {
+	t.until <- time.Now()
+	return t.r.Close()
+}
+
+func makeJournalFields(fields map[string]string) map[string]string {
+	result := make(map[string]string, len(fields))
+	for k, v := range fields {
+		result[fmt.Sprintf("__journal_%s", strings.ToLower(k))] = v
```

Comment:
prometheus standard for those are __labelname__ should we follow it ?

Reply:
(Using the journal field `PRIORITY` as an example) We could do it like that as `__priority__` or it could be in the style of Prometheus' meta labels (e.g., `__meta_kubernetes_pod_name`) as `__meta_journal_priority`. 

I think the meta label style matches the use case more closely here. What do you think?

Reply:
Yes I see what you mean thanks ! all good for me.